### PR TITLE
design of RNNOp

### PR DESCRIPTION
**Pull request by Superjom**
Fixes https://github.com/PaddlePaddle/Paddle/issues/3823

### CODE REVIEW

File: doc/design/ops/rnn.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+There are several important concepts:
+
+- stepnet, the network execute every time step 
+- memory, a variable
+- pre-memory, the value of the memory's value of the previous step time
+- init_memory, the variable to help initialize memory
+
+### step scopes
+Each RNN has more than one step times, and the stepnet will be executed in every step time.
```
Comment (wangkuiyi) [inline doc/design/ops/rnn.md]: The RNN might run one or more steps.

File: doc/design/ops/rnn.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+- stepnet, the network execute in every time step 
+- memory, a variable storing state in the current step, which is denoted as $h_t$.
+- pre-memory, the value of state in the previous time step, it can be denoted as $h_{t-1}$.
+- init-memory, the variable to help initialize state in the first time step.
+
+### step scopes
+
+<p aligh="center">
+<img src="./images/rnn.png"/><br/>
+fig 2 the RNN's data flow
```
Comment (wangkuiyi) [inline doc/design/ops/rnn.md]: fig 2 ==> Figure 2

File: doc/design/ops/rnn.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+<p aligh="center">
+<img src="./images/rnn.jpg"/>
+</p>
+
+The above diagram shows an RNN being unrolled into a full network.
+
+There are several important concepts:
+
+- stepnet, the network execute in every time step
```
Comment (wangkuiyi) [inline doc/design/ops/rnn.md]: stepnet => step-net

Comment (wangkuiyi) [inline doc/design/ops/rnn.md]: the network to be executed in each step

File: doc/design/ops/rnn.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+There are several important concepts:
+
+- stepnet, the network execute in every time step 
+- memory, a variable storing state in the current step, which is denoted as $h_t$.
+- pre-memory, the value of state in the previous time step, it can be denoted as $h_{t-1}$.
+- init-memory, the variable to help initialize state in the first time step.
+
+### step scopes
+
```
Comment (wangkuiyi) [inline doc/design/ops/rnn.md]: Step Scope

Comment (wangkuiyi) [inline doc/design/ops/rnn.md]: The step-net could have local variables defined. In each step of RNN execution, a scope is created to hold corresponding variables.  Such a scope is known as a *step scope*.

File: doc/design/ops/rnn.md [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+<p aligh="center">
+<img src="./images/rnn.png"/><br/>
+fig 2 the RNN's data flow
+</p>
+
+Each RNN might run one or more steps, each step runs the same step net.
```
Comment (wangkuiyi) [inline doc/design/ops/rnn.md]: All steps run the same step net.

File: doc/design/ops/rnn.md [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+We use `Scope` to store the contexts of all the step times:
```
Comment (wangkuiyi) [inline doc/design/ops/rnn.md]: Delete this paragraph.

File: doc/design/ops/rnn.md [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+- for each step, create a step Scope
```
Comment (wangkuiyi) [inline doc/design/ops/rnn.md]: Each step runs the following procedure:

1. create the step scope,
1. create local variables in the step scope, and
1. execute the step-net, which would use these variables.

File: doc/design/ops/rnn.md [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+- create all the temporary output variables in the Scope
+- execute the step-net, and each step will have its temporary outputs
+
+After all steps finished, RNNOp will collect the specific outputs of each step and merge them to a larger tensor.
```
Comment (wangkuiyi) [inline doc/design/ops/rnn.md]: After the execution of all steps, the RNNOp would compose its output from step outputs in those step scopes.

File: doc/design/ops/rnn.md [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+
+### memory and pre-memory
```
Comment (wangkuiyi) [inline doc/design/ops/rnn.md]: Memory and Ex-memory

File: doc/design/ops/rnn.md [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+a basic RNN is like:
```
Comment (wangkuiyi) [inline doc/design/ops/rnn.md]: An RNN step often has some status which needs to be passed to the next step.  These status are known as the memory, which is often referred by the step-net as variables.  A step often needs to refer to the memory value changed by the previous step.  We call it *ex-memory*.

Let's use a simply RNN as an example to explain memory and ex-memory:

File: doc/design/ops/rnn.md [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+
+$$
+h_t = U h_{t-1} + W x_t
+$$
+
+Here, $h_t$ is time $t$'s state, $h_t$ is time $t-1$'s state, in implementation, we call the a variable that store a state memory.
```
Comment (wangkuiyi) [inline doc/design/ops/rnn.md]: Here => where

Comment (wangkuiyi) [inline doc/design/ops/rnn.md]: where $h_t$ is the memory of step $t$, $h_{t-1}$ is the ex-memory, or the memory of step $t-1$.

Comment (wangkuiyi) [inline doc/design/ops/rnn.md]: ", in implementation, we call the a variable that store a state memory." can be deleted

File: doc/design/ops/rnn.md [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+In step time $t$, $h_t$ is memory, $h_{t-1}$ is pre-memory (short for previous memory).
```
Comment (wangkuiyi) [inline doc/design/ops/rnn.md]: "In step time $t$, $h_t$ is memory, $h_{t-1}$ is pre-memory (short for previous memory)." can be deleted

File: doc/design/ops/rnn.md [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+
+In each step scope
```
Comment (wangkuiyi) [inline doc/design/ops/rnn.md]: > +In each step scope
> +- each memory variable has a corresponding pre-memory variable
> +- before a time step executes, copy (or make a reference) the value of previous step scope's memory to the pre-memory variable in current step scope.

=>

> In the implementation, we can make an ex-memory variable either "refers to" the memory variable of the previous step, or copy the value of the previous memory variable to the current ex-memory variable.

File: doc/design/ops/rnn.md [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+
+- each memory variable has a corresponding pre-memory variable
+- before a time step executes, copy (or make a reference) the value of previous step scope's memory to the pre-memory variable in current step scope.
+
+### C++ API
```
Comment (wangkuiyi) [inline doc/design/ops/rnn.md]: The C++ API

File: doc/design/ops/rnn.md [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+- void CreateScopes(const framework::Scope& scope) const;
+  - create step scopes
+  - will be called both in InferShape and Run
+- void InitMemories(framework::Scope* step_scopes, bool infer_shape_mode) const;
+  - make a reference to the memory in previous step scope and memory in the current one.
+
+- void Run(const framework::Scope& scope, const platform::DeviceContext& dev_ctx) const;
+  - run all the time steps.
+
+### User interface
```
Comment (wangkuiyi) [inline doc/design/ops/rnn.md]: The Python Interface

File: doc/design/ops/rnn.md [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+X = some_op() # x is some operator's output, and is a LoDTensor
+a = some_op()
+
+# declare parameters
+W = pd.Variable(shape=[20, 30])
+U = pd.Variable(shape=[20, 30])
+
+rnn = pd.create_rnn_op(output_num=1)
+with rnn.stepnet():
+    x = rnn.add_input(X)
```
Comment (wangkuiyi) [inline doc/design/ops/rnn.md]: This example uses `rnn.add_input`. But the next example uses `rnn.segment_input`. Are they the same?

Comment (Superjom) [reply doc/design/ops/rnn.md]: yes, I will change all to `rnn.add_input`

Comment (Superjom) [reply doc/design/ops/rnn.md]: done

File: doc/design/ops/rnn.md [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+In the implementation, we can make an ex-memory variable either "refers to" the memory variable of the previous step,
+or copy the value of the previous memory value to the current ex-memory variable.
+
+### Usage in Python
+
+For more information on Block, please refer to the [design doc](https://github.com/PaddlePaddle/Paddle/blob/develop/doc/design/block.md).
+
+We can define an RNN's step-net using Block:
+
+```python
```
Comment (helinwang) [inline doc/design/ops/rnn.md]: Does this API works with attention model?

File: doc/design/ops/rnn.md [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+with rnn.stepnet():
+    x = rnn.add_input(X)
+    # declare a memory (rnn's step)
+    h = rnn.add_memory(init=a)
+    # h.pre_state() means previous memory of rnn
+    new_state = pd.add_two( pd.matmul(W, x) + pd.matmul(U, h.pre_state()))
+    # update current memory
+    h.update(new_state)
+    # indicate that h variables in all step scopes should be merged
+    rnn.set_output(0, h)
```
Comment (helinwang) [inline doc/design/ops/rnn.md]: What does "0" mean in `set_output`? Every `set_output` in this PR uses "0" as argument.

Comment (Superjom) [reply doc/design/ops/rnn.md]: 0 means the "0-th" argument

File: doc/design/ops/rnn.md (hunk 17 above)
Comment (Superjom) [reply doc/design/ops/rnn.md]: This syntax should be compatible with Paddle V1, but without the support of Beam Search.

File: doc/design/ops/rnn.md [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+    x: the input
+    '''
+    rnn = pd.create_rnn_op(output_num=1)
+    with rnn.stepnet():
+        sentence = rnn.add_input(paragraph, level=0)
+        h = rnn.add_memory(shape=[20, 30])
+        h.update(
+            pd.matmul(W, sentence) + pd.matmul(U, h.pre_state()))
+        # get the last state as sentence's info
+        rnn.set_output(0, h)
```
Comment (wangkuiyi) [inline doc/design/ops/rnn.md]: Is the `0` here indicating the first output?

How can we specify that an RNN should return just the output from the last step?

Comment (Superjom) [reply doc/design/ops/rnn.md]:
```python
rnn = pd.create_rnn_op()
with rnn.stepnet():
    x = rnn.set_inputs(X)
    # declare a memory (rnn's step)
    h = rnn.add_memory(init=a)
    # h.pre_state() means previous memory of rnn
    new_state = pd.add_two( pd.matmul(W, x) + pd.matmul(U, h.pre_state()))
    # update current memory
    h.update(new_state)
    # indicate that h variables in all step scopes should be merged
    rnn.set_outputs(h)

# output last step
out = rnn(output_all_steps=False)
```
can we use the argument `output_all_steps` to output all steps or just the last step?

File: doc/design/ops/rnn.md (hunk 16 above)
Comment (emailweixu) [reply doc/design/ops/rnn.md]: need to differentiate two types of input: sequence input and static input. Each instance has different static input. But for one instance, it's same across all time steps.

Comment (Superjom) [reply doc/design/ops/rnn.md]: the static inputs will be treated as global variables and that doesn't need to be passed as input.

the `add_input` statement only mark the sequence input that needs to be segmented for RNN's step times. @emailweixu

Comment (emailweixu) [reply doc/design/ops/rnn.md]: static inputs are different from parameters. They will still need to be splitted according to whether that instance is participating at a timestep, where parameters do not need to be splitted.

**Pull request merged**